### PR TITLE
Fix DRAM acceptance ratio

### PR DESCRIPTION
**Pull request by dawsoneliasen**

### CODE REVIEW

**Comment by bangerth** [comment]
There's a compile failure:
```
/home/runner/work/SampleFlow/SampleFlow/include/sampleflow/producers/delayed_rejection_mh.h:174:15: error: expected ‘;’ before ‘for’
               for (int j = 0; j < delay_stage; ++j)
               ^~~
/home/runner/work/SampleFlow/SampleFlow/include/sampleflow/producers/delayed_rejection_mh.h:174:31: error: ‘j’ was not declared in this scope
               for (int j = 0; j < delay_stage; ++j)
                               ^
/home/runner/work/SampleFlow/SampleFlow/include/sampleflow/producers/delayed_rejection_mh.h:174:31: note: suggested alternative: ‘jn’
               for (int j = 0; j < delay_stage; ++j)
                               ^
```

File: include/sampleflow/producers/delayed_rejection_mh.h [hunk 1]
```diff
@@ -165,26 +164,29 @@ namespace SampleFlow
...
-                                         proposal_distribution_ratio) * prev_acceptance_ratio;
-              double u = uniform_distribution(rng);
-              if (acceptance_ratio > 1 || acceptance_ratio >= u)
+              double orig_acceptance_ratio = likelihood_ratio * proposal_distribution_ratio;
+              double acceptance_ratio = orig_acceptance_ratio;
+              acceptance_ratios.push_back(orig_acceptance_ratio)
+
+              for (int j = 0; j < delay_stage; ++j)
+                {
+                  acceptance_ratio *= ((1 - acceptance_ratios[j + 1]) / (1 - acceptance_ratios[j]));
```
Comment (bangerth) [inline include/sampleflow/producers/delayed_rejection_mh.h:176]: It's hard to follow whether this is the correct formula. I think it would greatly help if you added the formula you're trying to implement as a comment, with the formula being what's listed in the paper.

**Comment by dawsoneliasen** [comment]
@bangerth One minor implementation trouble that I need some help with before I try to test... see my comment in the code. Other points of discussion: The alpha function is recursive, so it needs to be separated from `sample()`. C++ syntax-wise, did I correctly implement and call the instance method? Also, should we continue to expect the `perturb` function passed to the DRAM's `sample` method to return the proposal distribution ratio, even though we are now assuming that ratio to always be equal to 1 for delayed rejection sampling? Feel free to hold your thoughts for our meeting tomorrow.

File: include/sampleflow/producers/delayed_rejection_mh.h [hunk 2]
```diff
@@ -118,9 +126,53 @@ namespace SampleFlow
...
+        // in the case where no samples have been delayed, the acceptance ratio is calculated
+        // the same as regular MH; we assume that the proposal is symmetric, so the acceptance
+        // ratio is simply the likelihood ratio
+        if (i == 0)
+            return likelihood_ratio;
+        // otherwise, recursively compute acceptance ratio
+        double alpha = likelihood_ratio;
+        for (j = 0; j < y.size()) {
+            alpha *= (1 - alpha_fn(yi, y[i - j:i:-1], log_likelihood)) /
+                     (1 - alpha_fn(x, y[:j], log_likelihood));
```
Comment (dawsoneliasen) [inline include/sampleflow/producers/delayed_rejection_mh.h:170]: Here, I have the desired "slicing" behavior written in the python syntax that I'm familiar with. @bangerth , do you know an easy way to replicate this with C++ vectors? The first example, `y[i - j:i:-1]` means, take elements from `y`, starting at index `i - j` (inclusive), ending at index `i` (exclusive), and in reverse. The second example `y[:j]` just means take the first `j` elements.

Comment (bangerth) [reply include/sampleflow/producers/delayed_rejection_mh.h:162]: You could do something like this:
```suggestion
            alpha *= (1 - alpha_fn(yi, std::vector<double>(y.begin()+i - j, y.begin()+i]), log_likelihood)) /
                     (1 - alpha_fn(x, std::vector(y.begin(), y.begin()+j), log_likelihood));
```
but this doesn't reverse the first place. For that, you could think about using `std::vector::rbegin()/rend()`. You'd have to work out the correct `+i` or `+(i-1)` or `+(i-i)`, though.

File: include/sampleflow/producers/delayed_rejection_mh.h [hunk 3]
```diff
@@ -118,9 +118,57 @@ namespace SampleFlow
...
+              const std::vector<OutputType> &y,
+              const std::function<double (const OutputType &)> &log_likelihood)
+      {
+         // define the current delay stage as the number of rejected samples
+         // (number of proposed samples minus 1)
+         const int i = y.size() - 1;
+         // get the current proposed sample from the list of all proposed samples
+         const OutputType yi = y[i];
+         // get the likelihood ratio of the previous sample to the current proposed sample
+         const double likelihood_ratio = std::exp(log_likelihood(x) - log_likelihood(yi));
```
Comment (dawsoneliasen) [inline include/sampleflow/producers/delayed_rejection_mh.h:152]: Do not re-compute any `log_likelihood` values

File: include/sampleflow/producers/delayed_rejection_mh.h [hunk 4]
```diff
@@ -118,9 +118,57 @@ namespace SampleFlow
...
+    template <typename OutputType>
+    double 
+    DelayedRejectionMetropolisHastings<OutputType>::
+    alpha_fn (const OutputType &x,
+              const std::vector<OutputType> &y,
+              const std::function<double (const OutputType &)> &log_likelihood)
+      {
+         // define the current delay stage as the number of rejected samples
+         // (number of proposed samples minus 1)
+         const int i = y.size() - 1;
```
Comment (bangerth) [inline include/sampleflow/producers/delayed_rejection_mh.h:148]: give this a more descriptive name, say `n_rejected_samples`.

**Comment by dawsoneliasen** [comment]
@bangerth PTAL. The infinite recursion came from the fact that I was using 0 < j < y.size() instead of 0 < j < num_rejected_samples.

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
The `covariance_matrix_11` test does not actually use the DR sampler, so the reason why it fails must be unrelated to this pull request. Indeed, it also fails in #173 (which is a copy of #168). We will have to find out what is happening separately from this PR.

As a consequence, I'll merge as is!
